File: Java/Client/src/ResponseController.java
import java.io.IOException;
import java.util.NoSuchElementException;

public class ResponseController implements Runnable {

    static boolean serverHasRestarted;
    private final Client client;

    ResponseController(Client client) {
        this.client = client;
    }

    @Override
    @SuppressWarnings("InfiniteLoopStatement")
    public void run() {
        while (true) {
            try {
                String serverResponse = client.reader.nextLine();
                String[] serverResponseArr = serverResponse.split(":");
                boolean responseIsBroadcast = serverResponseArr[1].equals("broadcast");
                String response = serverResponseArr[0];
                if (responseIsBroadcast) {
                    client.broadcastQueue.add(response);
                } else {
                    client.responseQueue.add(response);
                }
            } catch (NoSuchElementException e) {
                System.err.println("Server killed");
                try {
                    if (Client.ballHolder == client.getPlayerId()) {
                        try {
                            ProcessBuilder pb = new ProcessBuilder("java", "-jar", ".\\lib\\BallGameServer.jar");
                            pb.start();
                            System.out.println("New server instance created");
                            Thread.sleep(2000);
                        } catch (IOException | InterruptedException exc) {
                            System.err.println(e.getMessage());
                        }
                    }
                    client.setUpConnections();
                    System.out.println("Joined new server instance");
                } catch (IOException ex) {
                    System.err.println("Error creating communication objects\n" + ex.getMessage());
                }
                serverHasRestarted = true;
            }
        }
    }
}

File: CS/Client/BallGameClient/ResponseController.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Threading;

namespace BallGameClient
{
    class ResponseController
    {
        public static bool serverHasRestarted;
        private Client client;

        public ResponseController(Client client)
        {
            this.client = client;
        }

        public void Run()
        {
            while (true)
            {
                try
                {
                    string serverResponse = client.reader.ReadLine();
                    if (serverResponse != null)
                    {
                        string[] serverResponseArr = serverResponse.Split(':');
                        bool responseIsBroadcast = serverResponseArr[1].Equals("broadcast");
                        string response = serverResponseArr[0];
                        if (responseIsBroadcast)
                        {
                            client.broadcastQueue.Add(response);
                        }
                        else
                        {
                            client.responseQueue.Add(response);
                        }
                    }
                }
                catch (IOException)
                {
                    Console.WriteLine("Server Killed");
                    if (Client.ballHolder == client.GetPlayerId())
                    {
                        // Start new server instance
                        Process newServerProcess = new Process();
                        newServerProcess.StartInfo.FileName = Directory.GetParent(Environment.CurrentDirectory).Parent.FullName + "\\lib\\BallGameServer.exe";
                        newServerProcess.StartInfo.CreateNoWindow = false;
                        newServerProcess.Start();
                        Console.WriteLine("New server instance created");
                        Thread.Sleep(2000);
                    }
                    client.SetUpConnections();
                    Console.WriteLine("Joined new server instance");
                    serverHasRestarted = true;
                    Console.WriteLine("Server restarted");
                }
            }
        }
    }
}
